src/ThermalShallowWaterExplicit.jl

```diff
@@ -0,0 +1,219 @@
+
+  # assemble the mass matrices
+  H1MM, RTMM, L2MM, H1MMchol, RTMMchol, L2MMchol = setup_and_factorize_mass_matrices(dΩ, R, S, U, V, P, Q)
+
+  # Project the initial conditions onto the trial spaces
+  b₁(q)   = ∫(q*h₀)dΩ
+  rhs1    = assemble_vector(b₁, Q)
+  hn      = FEFunction(Q, copy(rhs1))
+  ldiv!(L2MMchol, get_free_dof_values(hn))
+
+  b₂(v)   = ∫(v⋅u₀)dΩ
+  rhs2    = assemble_vector(b₂, V)
+  un      = FEFunction(V, copy(rhs2))
+  ldiv!(RTMMchol, get_free_dof_values(un))
+
+  b₃(s)   = ∫(s*f₀)*dΩ
+  rhs3    = assemble_vector(b₃, S)
+  f       = FEFunction(S, copy(rhs3))
+  ldiv!(H1MMchol, get_free_dof_values(f))
+
+  b₄(q)   = ∫(q*E₀)dΩ
+  rhs4    = assemble_vector(b₄, Q)
+  En      = FEFunction(Q, copy(rhs4))
+  ldiv!(L2MMchol, get_free_dof_values(En))
+
+  # work arrays
+  h_tmp = copy(get_free_dof_values(hn))
+  w_tmp = copy(get_free_dof_values(f))
+  # build the potential vorticity lhs operator once just to initialise
+  bmm(a,b) = ∫(a*hn*b)dΩ
+  H1h      = assemble_matrix(bmm, R, S)
+  H1hchol  = lu(H1h)
+
+  function run_simulation(pvd=nothing)
+    diagnostics_file = joinpath(output_dir,"tswe_diagnostics.csv")
+
+    hm1    = clone_fe_function(Q,hn)
+    hm2    = clone_fe_function(Q,hn)
+    hp     = clone_fe_function(Q,hn)
+    ϕ      = clone_fe_function(Q,hn)
+    Em1    = clone_fe_function(Q,En)
+    Em2    = clone_fe_function(Q,En)
+    Ep     = clone_fe_function(Q,En)
+
+    um1    = clone_fe_function(V,un)
+    um2    = clone_fe_function(V,un)
+    up     = clone_fe_function(V,un)
+    F      = clone_fe_function(V,un)
+    eF     = clone_fe_function(V,un)
+
+    wn     = clone_fe_function(S,f)
+    q1     = clone_fe_function(S,f)
+    q2     = clone_fe_function(S,f)
+    e1     = clone_fe_function(S,f)
+    e2     = clone_fe_function(S,f)
+
+    # first step, no leap frog integration
+    thermal_shallow_water_explicit_time_step!(hn, un, En, hp, up, Ep, ϕ, F, q1, q2, e1, e2, H1h, H1hchol, eF,
+                                              model, dΩ, dω, V, Q, R, S, f, hm1, um1, Em1, hm2, um2, Em2,
+                                              RTMMchol, L2MMchol, dt, τ, false)
+
+    if (write_diagnostics)
+      initialize_csv(diagnostics_file,"time", "mass", "vorticity", "kinetic", "internal", "power_k2p", "power_k2i")
+    end
+
+    if (write_diagnostics && write_diagnostics_freq==1)
+      compute_diagnostic_vorticity!(wn, dΩ, S, H1MMchol, un, get_normal_vector(model))
+      dump_diagnostics_thermal_shallow_water!(h_tmp, w_tmp,
+                                              model, dΩ, dω, S, L2MM, H1MM,
+                                              hn, un, En, wn, ϕ, F, eF, 1, dt,
+                                              diagnostics_file,
+                                              dump_diagnostics_on_screen)
+    end
+
+    # subsequent steps, do leap frog integration
+    # (now that we have the state at two previous time levels)
+    for istep in 2:N
```

Comment:
I am starting to get a bit worried about the explosion that we have in time stepping loops. The logic within the loop is almost verbatim for all solvers. See also https://github.com/gridapapps/GridapGeosciences.jl/pull/22#issuecomment-932746770. Perhaps at some point we could start thinking how to abstract this into a common piece of code reusable by all solvers so that we mitigate this issue.

Reply:
In any case, abstracting the thing into a common structure may come at loss of fliexibility to some extent if we do not design it with care/thinking about all possibilities. At this point, I am not sure if this is going to be possible/convenient. Writing my thoughs here ...

Reply:
Again, not sure. the Rosenbrock solver loops from 1:N not 2:N, and they all require different data, as mentioned above. With polymorphism I'm guessing it could be done (where each solver was a class with its own implementation of some virtual time_stepper and diagnostics, but doing this in a functional language I'm not sure...